File: android/app/src/main/kotlin/org/gaindrive/android/playback/wiim/WiiMEq.kt
package org.gaindrive.android.playback.wiim

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.put
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import org.gaindrive.android.net.SubsonicJson

/**
 * The equalizer state a WiiM reports.
 *
 * [preset] is null when the device would not say which one is loaded - see
 * [parseEqStat]. That is a real state and not an error: the switch is still
 * worth showing.
 */
data class WiiMEqState(
	val enabled: Boolean,
	val preset: String?,
	/**
	 * Fader positions in [WIIM_BANDS] order, or null when the device gave no
	 * usable set (the `EQGetStat` fallback, or a band missing or unreadable).
	 * Null is a real state too: the sheet then degrades to the switch and the
	 * preset list instead of drawing faders it could not fill.
	 */
	val bands: List<Int>? = null,
)

/**
 * Everything about the WiiM HTTP API that can be decided without a socket:
 * building a command URL and reading a response. `WiiMClient` is the I/O around
 * it.
 *
 * Split out for the reason `CastDeviceKind` and `LoadRetryWatcher` are: this
 * project has no ViewModel tests and no `MainDispatcherRule`, so what is worth
 * testing goes in a pure function. Everything here is exercised by `WiiMEqTest`.
 *
 * Bodies are read with safe casts rather than `@Serializable` DTOs, the house
 * rule for a payload from a device we do not control - the same reasoning as
 * `CastStatus.kt` and `CastProbe`. A malformed body must yield null, never an
 * exception: some of these calls run from a detached coroutine.
 */

/** The command a WiiM device answers on. Port 443, and always HTTPS. */
private const val PATH = "httpapi.asp"

/**
 * Builds the URL for one command.
 *
 * **The query must be assembled by `HttpUrl`, never concatenated.** Four of the
 * documented presets contain a space and one - `R&B` - contains an ampersand,
 * which unencoded would start a second query parameter and deliver the device
 * `EQLoad:R`. `addQueryParameter` percent-encodes both.
 *
 * Note the port: 443, not `CastDevice.port`. That field is 8009, the Cast v2
 * control port, and has nothing to do with this API.
 */
fun wiimUrl(address: String, command: String): HttpUrl =
	"https://$address/$PATH".toHttpUrl()
		.newBuilder()
		.addQueryParameter("command", command)
		.build()

/** The commands this app sends. `EQGetStat` is the fallback for [EQ_GET_BAND]. */
const val EQ_GET_BAND = "EQGetBand"
const val EQ_GET_STAT = "EQGetStat"
const val EQ_GET_LIST = "EQGetList"
const val EQ_ON = "EQOn"
const val EQ_OFF = "EQOff"

fun eqLoadCommand(preset: String): String = "EQLoad:$preset"

/**
 * The graphic EQ's ten fixed bands: `param_name` to fader label, in device
 * index order. Fixed rather than read from the device, unlike the preset list,
 * because `EQSetBand` must name them and a band this table does not know could
 * not be written anyway.
 */
val WIIM_BANDS: List<Pair<String, String>> = listOf(
	"band31hz" to "31",
	"band63hz" to "63",
	"band125hz" to "125",
	"band250hz" to "250",
	"band500hz" to "500",
	"band1khz" to "1k",
	"band2khz" to "2k",
	"band4khz" to "4k",
	"band8khz" to "8k",
	"band16khz" to "16k",
)

/**
 * The device's fader scale. How 0..99 maps to decibels is unverified (the WiiM
 * app draws +-12 dB), which is why the sheet shows offsets from flat and not
 * dB figures it cannot vouch for.
 */
const val WIIM_LEVEL_MIN = 0
const val WIIM_LEVEL_MAX = 99
const val WIIM_LEVEL_FLAT = 50

/**
 * The presets *HTTP API for WiiM Products v1.2* documents.
 *
 * Used only when `EQGetList` fails, so the sheet is never empty. It is not the
 * primary source because firmware versions differ and newer ones add presets -
 * a hardcoded list would silently hide them.
 */
val DOCUMENTED_PRESETS: List<String> = listOf(
	"Flat", "Acoustic", "Bass Booster", "Bass Reducer", "Classical", "Dance",
	"Deep", "Electronic", "Hip-Hop", "Jazz", "Latin", "Loudness", "Lounge",
	"Piano", "Pop", "R&B", "Rock", "Small Speakers", "Spoken Word",
	"Treble Booster", "Treble Reducer", "Vocal Booster",
)

/**
 * Reads an `EQGetBand` response, which carries the whole state at once:
 *
 *     {"status":"OK","source_name":"wifi","EQStat":"Off","Name":"Rock",
 *      "EQBand":[{"index":0,"param_name":"band31hz","value":71}, …]}
 *
 * This is why `EQGetBand` is the state read rather than `EQGetStat`: only this
 * one says *which* preset is loaded, and a picker that cannot show the current
 * selection is barely a picker. The band values feed the fader row; a body
 * without a usable set still yields the switch and the preset.
 */
fun parseEqBand(body: String): WiiMEqState? {
	val root = asObject(body) ?: return null
	val stat = root.string("EQStat") ?: return null
	return WiiMEqState(
		enabled = stat.equals("on", ignoreCase = true),
		preset = root.string("Name"),
		bands = parseBands(root),
	)
}

/**
 * The `EQBand` array as fader positions in [WIIM_BANDS] order.
 *
 * Matched by `param_name` rather than by the `index` field, so a reordered
 * array still reads correctly. All ten or nothing: a fader row with a hole in
 * it has no honest rendering, and half a curve written back would be a curve
 * the user never shaped.
 */
private fun parseBands(root: JsonObject): List<Int>? {
	val entries = (root["EQBand"] as? JsonArray)?.filterIsInstance<JsonObject>() ?: return null
	val byName = entries.mapNotNull { entry ->
		val name = entry.string("param_name") ?: return@mapNotNull null
		val value = (entry["value"] as? JsonPrimitive)?.intOrNull ?: return@mapNotNull null
		name to value
	}.toMap()
	return WIIM_BANDS.map { (name, _) ->
		(byName[name] ?: return null).coerceIn(WIIM_LEVEL_MIN, WIIM_LEVEL_MAX)
	}
}

/**
 * Reads an `EQGetStat` response - `{"EQStat":"On"}` or `{"EQStat":"Off"}`.
 *
 * The fallback. `EQGetBand` is documented by a third-party project rather than
 * by WiiM's own PDF, so a firmware without it is entirely possible; `EQGetStat`
 * is in the PDF. It costs the preset name, which is why it is second.
 */
fun parseEqStat(body: String): Boolean? =
	asObject(body)?.string("EQStat")?.equals("on", ignoreCase = true)

/**
 * Reads an `EQGetList` response: a JSON array of names, bare or wrapped.
 *
 * Served as `text/html` despite being JSON, which is why nothing here consults
 * the content type. An empty array is treated as a miss - a device with no
 * presets at all is not a state worth rendering, and the documented list is a
 * better answer than a blank sheet.
 *
 * **An array inside an object counts**, though no device has been seen sending
 * one: the firmware measured here answers this command with a bare array while
 * wrapping `EQGetBand` and every mutation in `{"status":"OK", …}`, so the shape
 * is plainly within its repertoire. Accepting both costs three lines, and the
 * consequence of refusing is silent - `presets()` never throws, so an unread
 * list is indistinguishable on screen from a device offering exactly the
 * documented set, and the difference only surfaces as a preset the owner made
 * themselves being absent. The field is found by shape and not by name because
 * nothing documents the name, and no such body carries a second array.
 */
fun parsePresets(body: String): List<String>? {
	val root = runCatching { SubsonicJson.parseToJsonElement(body) }.getOrNull() ?: return null
	val array = root as? JsonArray
		?: (root as? JsonObject)?.values?.firstNotNullOfOrNull { it as? JsonArray }
		?: return null
	val names = array.mapNotNull { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content }
		.filter { it.isNotBlank() }
	return names.ifEmpty { null }
}

/**
 * The command that writes all ten fader positions at once.
 *
 * Community-documented like `EQGetBand`, not in WiiM's own PDF; `WIIM.md`
 * records what is verified. Always the whole curve: the device accepts a
 * partial array, but sending one would make the outcome depend on state this
 * app last read rather than on what the user sees. Encoding the braces and
 * quotes into the URL is [wiimUrl]'s job, as everywhere else.
 */
fun eqSetBandCommand(levels: List<Int>): String {
	require(levels.size == WIIM_BANDS.size) {
		"expected ${WIIM_BANDS.size} band levels, got ${levels.size}"
	}
	val payload = buildJsonObject {
		put("EQBand", buildJsonArray {
			levels.forEachIndexed { index, value ->
				add(buildJsonObject {
					put("index", index)
					put("param_name", WIIM_BANDS[index].first)
					put("value", value.coerceIn(WIIM_LEVEL_MIN, WIIM_LEVEL_MAX))
				})
			}
		})
	}
	return "EQSetBand:$payload"
}

/**
 * Whether a command that answers `OK` or `Failed` succeeded.
 *
 * **Both documented shapes are accepted, because a device sends the other one.**
 * The PDF describes a plain-text `OK`, the OpenAPI description types the
 * response as an object with a `status` field, and this code believed the PDF -
 * so on real firmware every `EQLoad`, `EQOn` and `EQOff` was performed by the
 * device and reported to the user as "the device would not load that preset".
 * A command that is obeyed and then called a failure is the worst of the three
 * possible answers, so where the two documents disagree, accept both.
 *
 * `Failed` in either shape stays a failure: this is not "any 2xx is success".
 * Trimmed because a trailing newline is ordinary from this device.
 */
fun isOk(body: String): Boolean {
	val text = body.trim()
	if (text.equals("OK", ignoreCase = true)) return true
	return asObject(text)?.string("status")?.equals("OK", ignoreCase = true) == true
}

private fun asObject(body: String): JsonObject? =
	runCatching { SubsonicJson.parseToJsonElement(body) }.getOrNull() as? JsonObject

/** Safe by construction, like the rest of the foreign-JSON reading. */
private fun JsonObject.string(key: String): String? =
	(this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content?.takeIf { it.isNotBlank() }

File: src/wiimeq.cc
#include "wiimeq.hh"
#include "stamp.hh"
#include "textutil.hh"
#include "jsonread.hh"

#include <iostream>

#include <httplib.h>
#include <nlohmann/json.hpp>

namespace wiimeq {

const Band BANDS[10] = {
	{"band31hz",  "31"},
	{"band63hz",  "63"},
	{"band125hz", "125"},
	{"band250hz", "250"},
	{"band500hz", "500"},
	{"band1khz",  "1k"},
	{"band2khz",  "2k"},
	{"band4khz",  "4k"},
	{"band8khz",  "8k"},
	{"band16khz", "16k"},
	};

// One command, one reply body, or nullopt when the device did not answer.
static std::optional<std::string> command(const std::string& address,
                                          const std::string& cmd)
	{
	// Not CastDevice.port: 8009 is the Cast control port and has nothing to
	// do with this API.
	httplib::SSLClient cli(address, 443);
	// Every WiiM presents the same self-signed LinkPlay certificate
	// (CN=www.linkplay.com), which authenticates nothing: the user picked
	// this device off their own network, and nothing secret is sent to it.
	// Scoped to this client, never a global OpenSSL setting.
	cli.enable_server_certificate_verification(false);
	// A LAN round trip, on an httplib worker holding a browser request.
	cli.set_connection_timeout(3);
	cli.set_read_timeout(5);
	// The whole command is one query-parameter value and the device decodes
	// it as such. Encoding all of it is what keeps "R&B" from becoming
	// "EQLoad:R": an unencoded & ends the parameter.
	auto r = cli.Get(("/httpapi.asp?command=" + url_encode(cmd)).c_str());
	if (!r || r->status != 200) {
		std::cout << stamp() << "WiiM eq: " << cmd.substr(0, cmd.find(':'))
		          << " failed: "
		          << (r ? "HTTP " + std::to_string(r->status) : "no answer")
		          << std::endl;
		return std::nullopt;
		}
	// The device serves JSON as text/html; the content type means nothing.
	return r->body;
	}

// Success comes in two documented shapes, plain "OK" and {"status":"OK"},
// and where the documents differ, accept both: believing only one had every
// carried-out command reported as a failure (android/WIIM.md).
static bool is_ok(const std::string& body)
	{
	size_t a = body.find_first_not_of(" \t\r\n");
	if (a == std::string::npos) return false;
	size_t b = body.find_last_not_of(" \t\r\n");
	if (body.compare(a, b - a + 1, "OK") == 0) return true;
	auto j = nlohmann::json::parse(body, nullptr, false);
	return !j.is_discarded() && jstr(j, "status") == "OK";
	}

std::optional<State> state(const std::string& address)
	{
	// EQGetBand rather than EQGetStat: only this one names the loaded preset,
	// and a real Amp answers EQGetStat with {"status":"Failed"} anyway.
	auto body = command(address, "EQGetBand");
	if (!body) return std::nullopt;
	auto j = nlohmann::json::parse(*body, nullptr, false);
	if (j.is_discarded() || !j.is_object()) return std::nullopt;
	if (jstr(j, "status") == "Failed") return std::nullopt;

	State st;
	st.on     = jstr(j, "EQStat") == "On";
	st.preset = jstr(j, "Name");

	// Bands matched by param_name, all ten or nothing: a partial or garbled
	// array reads as "no faders" and the client degrades to the switch and
	// the preset list, while on/preset above stay valid.
	const auto& arr = jsub(j, "EQBand");
	if (arr.is_array()) {
		std::vector<int> vals(10, -1);
		for (const auto& e : arr) {
			std::string pn = jstr(e, "param_name");
			for (int i = 0; i < 10; i++)
				if (pn == BANDS[i].param) {
					int v = jint(e, "value");
					if (v >= LEVEL_MIN && v <= LEVEL_MAX) vals[i] = v;
					}
			}
		bool all = true;
		for (int v : vals) if (v < 0) all = false;
		if (all) st.bands = vals;
		}
	return st;
	}

std::optional<std::vector<std::string>> presets(const std::string& address)
	{
	auto body = command(address, "EQGetList");
	if (!body) return std::nullopt;
	auto j = nlohmann::json::parse(*body, nullptr, false);
	// A bare array on measured firmware; accept one wrapped in an object too,
	// because every other reply here is wrapped and the fallback (client-side)
	// hides its own failure.
	const nlohmann::json* arr = nullptr;
	if (j.is_array()) arr = &j;
	else if (j.is_object())
		for (const auto& [k, v] : j.items())
			if (v.is_array()) { arr = &v; break; }
	if (!arr) return std::nullopt;

	std::vector<std::string> names;
	for (const auto& e : *arr)
		if (e.is_string()) names.push_back(e.get<std::string>());
	if (names.empty()) return std::nullopt;
	return names;
	}

bool load_preset(const std::string& address, const std::string& name)
	{
	auto body = command(address, "EQLoad:" + name);
	if (!body || !is_ok(*body)) return false;
	// The documentation does not say whether EQLoad enables the equalizer;
	// sending EQOn makes the outcome the same either way, and picking "Rock"
	// while the switch is off must not be a silent no-op.
	return set_on(address, true);
	}

bool set_on(const std::string& address, bool on)
	{
	auto body = command(address, on ? "EQOn" : "EQOff");
	return body && is_ok(*body);
	}

bool set_bands(const std::string& address, const std::vector<int>& bands)
	{
	// Always the whole curve, so the outcome never depends on state the
	// caller last read rather than on what the user sees.
	if (bands.size() != 10) return false;
	for (int v : bands)
		if (v < LEVEL_MIN || v > LEVEL_MAX) return false;
	nlohmann::json arr = nlohmann::json::array();
	for (int i = 0; i < 10; i++)
		arr.push_back({{"index", i}, {"param_name", BANDS[i].param},
		               {"value", bands[i]}});
	auto body = command(address,
	                    "EQSetBand:" + nlohmann::json{{"EQBand", arr}}.dump());
	return body && is_ok(*body);
	}

	}
